tiledb/sm/storage_manager/storage_manager.cc

```diff
@@ -198,325 +138,193 @@ std::tuple<
 
-  if (!st.ok()) {
-    open_array->mtx_unlock();
-    array_close_for_reads(array_uri);
-    return {st, std::nullopt, std::nullopt, std::nullopt};
-  }
+  return {
+      Status::Ok(), array_schema_latest, array_schemas_all, fragment_metadata};
+}
 
-  // Unlock the array mutex
-  open_array->mtx_unlock();
+std::tuple<
+    Status,
+    std::optional<ArraySchema*>,
+    std::optional<std::unordered_map<std::string, tdb_shared_ptr<ArraySchema>>>,
+    std::optional<std::vector<tdb_shared_ptr<FragmentMetadata>>>>
+StorageManager::array_open_for_reads(Array* array) {
+  auto timer_se = stats_->start_timer("array_open_for_reads");
+  auto&& [st, array_schema_latest, array_schemas_all, fragment_metadata] =
+      load_array_schemas_and_fragment_metadata(
+          array->array_uri(),
+          array->memory_tracker(),
+          *array->encryption_key(),
+          array->timestamp_start(),
+          array->timestamp_end_opened_at());
+  RETURN_NOT_OK_TUPLE(st, std::nullopt, std::nullopt, std::nullopt);
 
-  return {Status::Ok(),
-          open_array->array_schema_latest(),
-          open_array->array_schemas_all(),
-          fragment_metadata};
+  // Mark the array as open
+  std::lock_guard<std::mutex> lock{open_arrays_mtx_};
+  open_arrays_.insert(array);
+
+  return {
+      Status::Ok(), array_schema_latest, array_schemas_all, fragment_metadata};
 }
 
 std::tuple<
     Status,
     std::optional<ArraySchema*>,
     std::optional<std::unordered_map<std::string, tdb_shared_ptr<ArraySchema>>>>
-StorageManager::array_open_for_reads_without_fragments(
-    const URI& array_uri, const EncryptionKey& enc_key) {
-  auto timer_se = stats_->start_timer("read_array_open");
+StorageManager::array_open_for_reads_without_fragments(Array* array) {
+  auto timer_se = stats_->start_timer("array_open_for_reads_without_fragments");
 
-  auto open_array = (OpenArray*)nullptr;
-  Status st = array_open_without_fragments(array_uri, enc_key, &open_array);
-  if (!st.ok()) {
-    return {st, std::nullopt, std::nullopt};
-  }
+  // Load array schemas
+  auto&& [st_schemas, array_schema_latest, array_schemas_all] =
+      load_array_schemas(array->array_uri(), *array->encryption_key());
+  RETURN_NOT_OK_TUPLE(st_schemas, std::nullopt, std::nullopt);
 
-  // Unlock the array mutex
-  open_array->mtx_unlock();
+  // Mark the array as now open
+  std::lock_guard<std::mutex> lock{open_arrays_mtx_};
+  open_arrays_.insert(array);
 
-  return {Status::Ok(),
-          open_array->array_schema_latest(),
-          open_array->array_schemas_all()};
+  return {Status::Ok(), array_schema_latest, array_schemas_all};
 }
 
 std::tuple<
     Status,
     std::optional<ArraySchema*>,
     std::optional<std::unordered_map<std::string, tdb_shared_ptr<ArraySchema>>>>
-StorageManager::array_open_for_writes(
-    const URI& array_uri, const EncryptionKey& encryption_key) {
-  if (!vfs_->supports_uri_scheme(array_uri))
+StorageManager::array_open_for_writes(Array* array) {
```

Comment:
Could we move this to a tdb_shared_ptr?

Reply:
Not here, as it will require changing this everywhere. This is my next PR.